src/main/java/net/neoforged/neoforge/registries/BaseMappedRegistry.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.registries;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.registries.callback.AddCallback;
+import net.neoforged.neoforge.registries.callback.BakeCallback;
+import net.neoforged.neoforge.registries.callback.ClearCallback;
+import net.neoforged.neoforge.registries.callback.RegistryCallback;
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.Internal
+public abstract class BaseMappedRegistry<T> implements Registry<T> {
+    protected final List<AddCallback<T>> addCallbacks = new ArrayList<>();
+    protected final List<BakeCallback<T>> bakeCallbacks = new ArrayList<>();
+    protected final List<ClearCallback<T>> clearCallbacks = new ArrayList<>();
+    final Map<ResourceLocation, ResourceLocation> aliases = new HashMap<>();
+    private int maxId = Integer.MAX_VALUE - 1;
+    private boolean sync;
+
+    void setSync(boolean sync) {
+        this.sync = sync;
+    }
+
+    @Override
+    public boolean doesSync() {
+        return this.sync;
+    }
+
+    void setMaxId(int maxId) {
+        this.maxId = maxId;
+    }
+
+    @Override
+    public int getMaxId() {
+        return this.maxId;
+    }
+
+    @Override
+    public void addCallback(RegistryCallback<T> callback) {
+        if (callback instanceof AddCallback<T> addCallback)
+            this.addCallbacks.add(addCallback);
+        if (callback instanceof BakeCallback<T> bakeCallback)
+            this.bakeCallbacks.add(bakeCallback);
+        if (callback instanceof ClearCallback<T> clearCallback)
+            this.clearCallbacks.add(clearCallback);
+    }
+
+    @Override
+    public void addAlias(ResourceLocation from, ResourceLocation to) {
+        if (from.equals(to))
+            return;
+        if (this.aliases.containsKey(from)) {
+            ResourceLocation old = this.aliases.get(from);
+            if (!old.equals(to))
+                throw new IllegalStateException("Duplicate alias with key \"" + from + "\" attempting to map to \"" + to + "\", found existing mapping \"" + old + "\"");
+        }
+        if (resolve(from).equals(to))
+            throw new IllegalStateException("Infinite alias loop detected: from " + from + " to " + to);
+        this.aliases.put(from, to);
+    }
+
+    @Override
+    public ResourceLocation resolve(ResourceLocation name) {
+        if (this.containsKey(name))
+            return name;
+
+        ResourceLocation alias = this.aliases.get(name);
+        if (alias == null)
+            return name;
+
+        return resolve(alias);
+    }
+
+    @Override
+    public ResourceKey<T> resolve(ResourceKey<T> key) {
+        return ResourceKey.create(this.key(), resolve(key.location()));
```

Comment:
Should we consider optimizing this a bit for the likely case that `resolve(ResourceLocation)` returns the same value? As in, `if (resolve(key.location()) == key.location()) return key;` (with variables of course to reduce the calls needed).

Reply:
Fixed in f2801cb37.